### PR TITLE
Bar Chart Selection

### PR DESCRIPTION
**Pull request by dvzacharycutler**
Closes https://github.com/datavisyn/reprovisyn/issues/294

Changes for allowing clickable bar charts. All Segments of grouped bar charts and different charts in faceted charts are also clickable. Currently, a bar is highlighted only if every point in that bar is selected. Outside selections can also affect this. 

https://user-images.githubusercontent.com/87644761/171185018-898b311b-a665-47b8-bc3e-32212f31344e.mp4

CTRL allows clicking of multiple bars at once

https://user-images.githubusercontent.com/87644761/171185067-1e071cf8-d562-4bd2-8050-1cb52b1ebeaa.mp4

There are still two problems I don't really know how to address. 

First, plotly seems to be finicky about click events and sometimes thinks they're drag events instead so a bar doesn't get selected when you think you clicked. I turned off dragmode, but this seems to still occur. Not super bad, but annoying.

Second, the expected interaction for me is that if I click on white space in my plot I would deselect any bar segments. But, for some unknown reason, plotly doesn't send me events unless my click is on an actual bar, and I haven't figured a way around this yet (I could, of course, select the g layer manually and add an on click but... gross). 

So, I'm going to adopt the strategy of Cato in ancient Rome and end this PR saying, as always, I think we should rewrite this component custom without plotly.

### CODE REVIEW

**Review by thinkh — CHANGES_REQUESTED** [review]
Thanks for the enhancement of the bar chart. I think the functionality is really important to the user.

Please see my comments below. Regarding your points:

> First, plotly seems to be finicky about click events and sometimes thinks they're drag events instead so a bar doesn't get selected when you think you clicked. I turned off dragmode, but this seems to still occur. Not super bad, but annoying.

I'm not exactly sure what you mean. Is it possible to record a screencast showing this behavior?

> Second, the expected interaction for me is that if I click on white space in my plot I would deselect any bar segments. But, for some unknown reason, plotly doesn't send me events unless my click is on an actual bar, and I haven't figured a way around this yet (I could, of course, select the g layer manually and add an on click but... gross).

Did you check the following solutions:

* https://community.plotly.com/t/unselect-points-when-click-empty-area/3262/3
* https://github.com/plotly/plotly.js/issues/2696

> So, I'm going to adopt the strategy of Cato in ancient Rome and end this PR saying, as always, I think we should rewrite this component custom without plotly.

I like your humor. 😄

File: src/vis/bar/utils.ts [hunk 1]
```diff
@@ -188,6 +200,17 @@ async function setPlotsWithGroupsAndMultiples(
           marker: {
             color: scales.color(uniqueGroup),
           },
+          // @ts-ignore
```
[the remark this answers is not in the document]
Comment (dvzacharycutler) [reply src/vis/bar/utils.ts:203]: I think plotlys types are just wrong, but its also possible something else in the config magically changes the type to make it think it cant have a selected prop. But it does have an unselected prop. So going to go with something wrong with plotly.

File: src/vis/bar/BarVis.tsx [hunk 2]
```diff
@@ -75,6 +80,54 @@ export function BarVis({
...
+  const finalTraces = useMemo(() => {
+    if (!traces) {
+      return null;
+    }
+
+    let selectedFlag = false;
+
+    const editedTraces = { ...traces };
+
+    const allSelected = [];
```
[the remark this answers is not in the document]
Comment (dvzacharycutler) [reply src/vis/bar/BarVis.tsx:93]: Agreed, though after looking again I can just remove this array. Was an artifact of doing it a different way.

File: src/vis/bar/BarVis.tsx [hunk 3]
```diff
@@ -75,6 +80,54 @@ export function BarVis({
...
+    editedTraces?.plots.forEach((plot) => {
+      const tracePoints = plot.data.customdata;
+
+      const selectedArr = [];
```
[the remark this answers is not in the document]
Comment (dvzacharycutler) [reply src/vis/bar/BarVis.tsx:97]: Done

File: src/vis/bar/BarVis.tsx [hunk 4]
```diff
@@ -75,6 +80,54 @@ export function BarVis({
 
   const { value: traces, status: traceStatus, error: traceError } = useAsync(createBarTraces, [columns, config, scales]);
 
+  // Make sure selected values is right for each plot.
+  const finalTraces = useMemo(() => {
+    if (!traces) {
+      return null;
+    }
+
+    let selectedFlag = false;
```
[the remark this answers is not in the document]
Comment (dvzacharycutler) [reply src/vis/bar/BarVis.tsx:89]: Done

File: src/vis/bar/BarVis.tsx [hunk 5]
```diff
@@ -136,25 +198,50 @@ export function BarVis({
...
+            data={traceData}
             layout={layout}
             config={{ responsive: true, displayModeBar: false }}
             useResizeHandler
             style={{ width: '100%', height: '100%' }}
+            // The types on this event dont seem to work correctly with Plotly types, thus the any typing.
+            onClick={(e: any) => {
+              const selectedPoints: string[] = e.points[0].customdata;
+
+              console.log(e);
```
[the remark this answers is not in the document]
Comment (dvzacharycutler) [reply src/vis/bar/BarVis.tsx:213]: Done

File: src/vis/bar/utils.ts [hunk 6]
```diff
@@ -1,3 +1,5 @@
+/* eslint-disable @typescript-eslint/ban-ts-comment */
+/* eslint-disable @typescript-eslint/naming-convention */
```
[the remark this answers is not in the document]
Comment (dvzacharycutler) [reply src/vis/bar/utils.ts:2]: no we shouldnt need them, removed

File: src/vis/sidebar/AggregateTypeSelect.tsx [hunk 7]
```diff
@@ -13,16 +12,6 @@ interface AggregateTypeSelectProps {
...
 
-function formatOptionLabel(option, ctx) {
-  return (
-    <>
-      <Highlighter searchWords={[ctx.inputValue]} autoEscape textToHighlight={option.name} />
-      {option.description && <span className="small text-muted ms-1">{option.description}</span>}
-      {option.disabled ? <i className="ms-1 fas fa-question-circle" title="No numerical columns available for this aggregation type" /> : null}
-    </>
-  );
-}
```
[the remark this answers is not in the document]
Comment (dvzacharycutler) [reply src/vis/sidebar/AggregateTypeSelect.tsx:24]: the format option label moved into the sidebar/utils.tsx, so that I didnt duplicate it in every one of the react-select components. This one was just a leftover that wasnt being used anyways.

File: src/vis/bar/utils.ts [hunk 8]
```diff
@@ -273,6 +316,17 @@ async function setPlotsWithGroups(
...
+  // const hasSelected = plots.find((plot) => plot.data.selectedpoints !== null);
+
+  // // Need to check if we need to tell all of the traces that there is a selection so they can be opaque
+  // if (hasSelected) {
+  //   plots.forEach((p) => {
+  //     if (p.data.selectedpoints === null) {
+  //       p.data.selectedpoints = [];
+  //     }
+  //   });
+  // }
```
[the remark this answers is not in the document]
Comment (dvzacharycutler) [reply src/vis/bar/utils.ts:328]: Done

File: src/vis/bar/BarVis.tsx [hunk 9]
```diff
@@ -136,25 +198,50 @@ export function BarVis({
...
+
+              let removeSelectionFlag = true;
+
+              for (const j of selectedPoints) {
```
[the remark this answers is not in the document]
Comment (dvzacharycutler) [reply src/vis/bar/BarVis.tsx:217]: Did some typing to get around the incorrect plotly types so we had typing here, and renamed variable to pointId

File: src/vis/bar/utils.ts [hunk 10]
```diff
@@ -172,11 +175,20 @@ async function setPlotsWithGroupsAndMultiples(
 
       const plotAggregateAxisName = createAxisLabel(aggregateType, aggregateColumn);
 
+      let valIdArr = uniqueColVals.map((val) => []);
```
[the remark this answers is not in the document]
Comment (dvzacharycutler) [reply src/vis/bar/utils.ts:178]: Added typings and a comment

File: src/vis/bar/utils.ts [hunk 11]
```diff
@@ -172,11 +175,20 @@ async function setPlotsWithGroupsAndMultiples(
...
+
+      catColValues.resolvedValues.forEach((row) => valIdArr[uniqueColVals.indexOf(row.val as string)].push(row.id));
+
+      valIdArr = valIdArr.map((arr) => arr.filter((val) => allGroupObjsIds.has(val) && allMultiplesObjsIds.has(val)));
```
[the remark this answers is not in the document]
Comment (dvzacharycutler) [reply src/vis/bar/utils.ts:182]: Yes, totally agree with you the data wrangling logic in this file is out of hand. And also happens with very minor changes like four times. I've mentioned it before, but I'm not really sure what to do about it. I've never seen these libraries before, but they do look helpful! Definitely seems worth a look.

File: src/vis/bar/utils.ts [hunk 12]
```diff
@@ -308,19 +362,38 @@ async function setPlotsWithMultiples(
       .flat();
     const plotAggregateAxisName = createAxisLabel(aggregateType, aggregateColumn);
 
+    let valIdArr = uniqueColVals.map((val) => []);
```
[the remark this answers is not in the document]
Comment (dvzacharycutler) [reply src/vis/bar/utils.ts:365]: Done

File: src/vis/bar/BarVis.tsx [hunk 13]
```diff
@@ -75,6 +80,54 @@ export function BarVis({
...
+    const allSelected = [];
+    editedTraces?.plots.forEach((plot) => {
+      const tracePoints = plot.data.customdata;
+
+      const selectedArr = [];
+      tracePoints.forEach((trace, index) => {
+        if (trace.length === 0 || selectedList.length < trace.length) {
+          return;
+        }
+        for (const i of trace) {
```
[the remark this answers is not in the document]
Comment (dvzacharycutler) [reply src/vis/bar/BarVis.tsx:102]: Did some variable renaming and the above changes also gave us typing here.

I've mentioned this before, but the difficulty here is that our incoming selection is just a list of ids, we have no concept of a selected category. So, we have to just brute force check if an entire section of a bar is selected in order for this to work. I don't think there's any easy way to fix/flatten this, and at some point id consider adding a concept of a selected category, especially in reprovisyn. It's linear worst case, though should be much faster if not many items are selected.

File: src/vis/bar/BarVis.tsx [hunk 14]
```diff
@@ -65,6 +67,9 @@ export function BarVis({
   columns,
   setConfig,
   scales,
+  selectionCallback = () => null,
+  selectedMap = {},
+  selectedList = [],
```
[the remark this answers is not in the document]
Comment (dvzacharycutler) [reply src/vis/bar/BarVis.tsx:72]: More or less yes, selectedList is what gets passed into the vis component and we internally create selectedMap to help with speed stuff. We pass them both around frequently since they are used in different ways, but we could just pass around selectedMap and infer the list as you said. But this seems slightly more confusing to me than just using the list that we already have?

Added types for the map

**Comment by dvzacharycutler** [comment]
> I'm not exactly sure what you mean. Is it possible to record a screencast showing this behavior?

It's kind of hard to see since you cant tell when I click, but the second and last time that I click here nothing happens. The line in plotly flashes, so it definitely knows something happened, but no event gets sent. It occurs if you try and click after quickly moving your mouse. 

https://user-images.githubusercontent.com/87644761/173055999-1ce46ed4-e1a5-40a2-bf1d-5d6c3d35ed51.mp4

> Did you check the following solutions:

The first one there isn't quite the problem, though it seems like the second one is! Unfortunately its a feature that as far as i can tell is still pending, after four years :)

File: src/vis/bar/BarVis.tsx [hunk 15]
```diff
@@ -75,6 +80,54 @@ export function BarVis({
...
+      return null;
+    }
+
+    let selectedFlag = false;
+
+    const editedTraces = { ...traces };
+
+    const allSelected = [];
+    editedTraces?.plots.forEach((plot) => {
+      const tracePoints = plot.data.customdata;
```
[the remark this answers is not in the document]
Comment (thinkh) [reply src/vis/bar/BarVis.tsx]: 👍

File: src/vis/bar/BarVis.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (thinkh) [reply src/vis/bar/BarVis.tsx]: Ok, I see. Thanks for the explanation.

File: src/vis/bar/utils.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (thinkh) [reply src/vis/bar/utils.ts:181]: We should try to find some time for a refactoring. Because even now, with some time in between, the data wrangling code is hard to follow and read between all the loops. We can also add some unit tests at least for the utils and/or break the large functions up in smaller pieces to make it testable/readable/maintainable.

File: src/vis/sidebar/AggregateTypeSelect.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (thinkh) [reply src/vis/sidebar/AggregateTypeSelect.tsx:24]: Sounds good.

File: src/vis/bar/BarVis.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (thinkh) [reply src/vis/bar/BarVis.tsx:72]: Ok, then we leave it like this for now.

**Review by thinkh — APPROVED** [review]
Thanks for addressing my comments. It looks now as good as it can get. I will open a follow-up issue to refactor the data wrangling part and write some unit tests. 

Regarding the click: I carefully watched your video multiple times and also tested it myself locally. However, I could not reproduce a bug here (or I still don't know what to look exactly for). Hence, I suggest to merge this PR and open another issue if we have a real bug there. But thanks for mentioning this as part of the initial PR conversation.
